### PR TITLE
Enable s2i windows testing

### PR DESCRIPTION
**Pull request by jim-minter**
- add 'build-sti-base-windows' vagrant command to build windows AMI with required prereqs
- hooks to enable simplification of Vagrantfile (auto-discover AMI id, auto-configure standard Vagrantfile parameters for AWS)
- hook to wait for Windows instance to report ready after sysprep
- provisioners to allow a Windows instance to be autoconfigured to reasonably securely use Docker on a helper RHEL instance
- miscellaneous fixes

@bparees 
@stevekuznetsov

### CODE REVIEW

File: lib/vagrant-openshift/action/build_sti_base_windows.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+          @env[:ui].info("Using base AMI #{images.last.name}, #{images.last.id}")
+
+          @env[:ui].info("Creating instance...")
+          instance = compute.servers.create(
+            :tags => {"Name" => "#{@options[:instance_prefix]}-windows2012r2_#{Time.now.to_i}"},
+            :image_id => images.last.id,
+            :flavor_id => "m4.large",
+            :key_name => "libra",
+            :subnet_id => "subnet-cf57c596",
+            :user_data => <<'USERDATA'
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/build_sti_base_windows.rb]: In commands like `build-origin-base` we've moved to having script files in this repo that get deposited on the remote system and executed explicitly. I find that it is a lot less mentally taxing to not have to consider all of the shell script syntax in the context of being an inline Ruby string, etc. If it's a valid approach in this instance, I think it would be a better choice for long-term maintainability of the provision script.

File: lib/vagrant-openshift/action/download_artifacts_sti.rb [hunk 2]
```diff
@@ -34,11 +34,15 @@ def call(env)
 
           artifacts_dir = Pathname.new(File.expand_path(machine.env.root_path + "artifacts"))
           download_map = {
-            "/var/log/yum.log"                     => artifacts_dir + "yum.log",
-            "/var/log/secure"                      => artifacts_dir + "secure",
-            "/var/log/audit/audit.log"             => artifacts_dir + "audit.log",
             "/tmp/sti/"                            => artifacts_dir + "sti/"
           }
+          if machine.config.ssh.username != "Administrator"
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/download_artifacts_sti.rb]: I'm assuming this is detecting Windows? The `if` condition could be a little more clear.

File: lib/vagrant-openshift/action/run_sti_tests.rb [hunk 3]
```diff
@@ -52,7 +52,7 @@ def call(env)
           _,_,env[:test_exit_code] = sudo(env[:machine], %{
 set -e
 pushd #{Constants.build_dir}/source-to-image >/dev/null
-export PATH=/data/src/github.com/openshift/source-to-image/_output/local/go/bin:/data/src/github.com/openshift/source-to-image/_output/local/bin/linux/amd64:$PATH
+export PATH=/data/src/github.com/openshift/source-to-image/_output/local/go/bin:/data/src/github.com/openshift/source-to-image/_output/local/bin/linux/amd64:/data/src/github.com/openshift/source-to-image/_output/local/bin/windows/amd64:$PATH
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/run_sti_tests.rb]: Wouldn't someone running these tests locally on their own machine need the `$PATH` augment as well? Why not have the `$PATH` change inside of the test script?

File: lib/vagrant-openshift/helper/command_helper.rb [hunk 4]
```diff
@@ -235,6 +235,13 @@ def bash_systemd_polling_functions
         return command
       end
 
+      def get_ssh_group(ssh_user)
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/helper/command_helper.rb]: If we have access to the machine, why not actually determine the group for the user?

File: lib/vagrant-openshift/hooks/configure_aws.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+              box_info = YAML.load_file(Pathname.new(File.expand_path("#{__FILE__}/../../templates/command/init-openshift/box_info.yaml")))
+              if !aws.ami.start_with?("ami-")
+                os, stage = aws.ami.split(":", 2)
+                env[:machine].config.ssh.username = box_info[os.to_sym][stage.to_sym][:aws][:ssh_user]
+              end
+            end
+            env[:machine].config.ssh.private_key_path = [aws_creds["AWSPrivateKeyPath"]]
+            env[:machine].config.vm.box = "https://github.com/mitchellh/vagrant-aws/raw/master/dummy.box"
+            env[:machine].config.vm.synced_folders["/vagrant"] = {:disabled => true, :guestpath => "/vagrant", :hostpath => "."}
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: `guestpath` and `hostpath` are ignored since we disable this, right?

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+openssl req -new -key /tmp/ca.key -subj /CN=CA -x509 -sha1 -days 7 -set_serial 1 -out /tmp/ca.crt
+openssl genrsa -out /tmp/server.key
+openssl req -new -key /tmp/server.key -subj /CN=$IP -out /tmp/server.csr
+echo subjectAltName=IP:$IP | openssl x509 -req -in /tmp/server.csr -CA /tmp/ca.crt -CAkey /tmp/ca.key -days 7 -extfile /dev/stdin -set_serial 2 -out /tmp/server.crt
+rm -f /tmp/server.csr
+openssl genrsa -out /tmp/client.key
+openssl req -new -key /tmp/client.key -subj /CN=client -out /tmp/client.csr
+echo extendedKeyUsage=clientAuth | openssl x509 -req -in /tmp/client.csr -CA /tmp/ca.crt -CAkey /tmp/ca.key -days 7 -extfile /dev/stdin -set_serial 3 -out /tmp/client.crt
+rm -f /tmp/client.csr
+grep -q tlsverify /etc/sysconfig/docker || sed -i -e "s|^OPTIONS='|OPTIONS='--tlsverify --tlscacert=/tmp/ca.crt --tlscert=/tmp/server.crt --tlskey=/tmp/server.key -H unix:///var/run/docker.sock -H tcp://0.0.0.0:2376 |" /etc/sysconfig/docker
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: Would prefer to use proper `if/then` statements:
```sh
if ! grep -q 'tlsverify' /etc/sysconfig/docker; then
    # whatever
fi
```

Comment (stevekuznetsov) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: Won't this clobber whatever is in `OPTIONS=`?

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+systemctl restart docker.service
+echo -e "export DOCKER_HOST=tcp://$IP:2376\nexport DOCKER_TLS_VERIFY=1" >/tmp/dockerenv
+}
+          @machine.communicate.sudo(cmd)
+          @machine.communicate.download("/tmp/ca.crt", ".vagrant/docker-config/ca.crt")
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: @danmcp these are going to be left on the Jenkins box in the workspace, do we need to worry about cleaning them up?

File: lib/vagrant-openshift/hooks/find_ami.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      class FindAMI
+        def initialize(app, env)
+          @app = app
+        end
+
+        def call(env)
+          if env[:machine].config.openshift.autoconfigure_aws == true
+            aws = env[:machine].config.vm.get_provider_config(:aws)
+            if !aws.nil? && !aws.ami.nil? && !aws.ami.start_with?("ami-")
+              box_info = YAML.load_file(Pathname.new(File.expand_path("#{__FILE__}/../../templates/command/init-openshift/box_info.yaml")))
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/find_ami.rb]: We already did this work when we created `.vagrant-openshift.json` -- I can't remember, are you also creating entries in that file?

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb (hunk 7 above)
Comment (danmcp) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: They are files generated during the run right?

File: lib/vagrant-openshift/hooks/configure_aws.rb (hunk 5 above)
Comment (jim-minter) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: Yes; the reason I left them in is because of the recommendation for disabling the default synced folder at https://www.vagrantup.com/docs/synced-folders/basic_usage.html#disabling :

Adding `config.vm.synced_folder ".", "/vagrant", disabled: true` to a Vagrantfile as recommended at the above location results in the hash `{:disabled => true, :guestpath => "/vagrant", :hostpath => "."}` being stored, rather just than `{:disabled => true}`.  I figured, this being Vagrant, that not shortcutting may be less likely to cause an issue somewhere else.

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb (hunk 6 above)
Comment (jim-minter) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: To the latter point: no, this specifically prepends to the OPTIONS string.

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb (hunk 7 above)
Comment (jim-minter) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: Yes, they are generated when the provisioner is run.

File: lib/vagrant-openshift/action/build_sti_base_windows.rb (hunk 1 above)
Comment (jim-minter) [inline lib/vagrant-openshift/action/build_sti_base_windows.rb]: Because this is a user data script run on an AMI we don't control, `@env[:machine].communicate.upload` (the pattern used elsewhere for resources files) is not available here.  The alternative is to open a resources file client-side and assign its contents to `:user_data`.  Given that this string is already not interpolated or escaped, do you still want me to make that change?

File: lib/vagrant-openshift/action/download_artifacts_sti.rb (hunk 2 above)
Comment (jim-minter) [inline lib/vagrant-openshift/action/download_artifacts_sti.rb]: Yes; sorry - how do you mean more clear?

File: lib/vagrant-openshift/action/run_sti_tests.rb (hunk 3 above)
Comment (jim-minter) [inline lib/vagrant-openshift/action/run_sti_tests.rb]: Agreed - will revert this change, update s2i and add a comment indicating that the entire PATH line can be removed once the s2i change merges.

File: lib/vagrant-openshift/helper/command_helper.rb (hunk 4 above)
Comment (jim-minter) [inline lib/vagrant-openshift/helper/command_helper.rb]: It's feasible, but pretty much everywhere in this code base assumes ssh_group == ssh_user; actually determining it adds another RTT of latency and more scope for a network glitch.  It's tradition on Linux systems that users have an identically named group; it's tradition on Windows/Cygwin that Administrator's group is Administrators.

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb (hunk 6 above)
Comment (jim-minter) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: Will change to proper `if/then` statements.

File: lib/vagrant-openshift/hooks/find_ami.rb (hunk 8 above)
Comment (jim-minter) [inline lib/vagrant-openshift/hooks/find_ami.rb]: No, I'm specifically not touching `.vagrant-openshift.json` because it doesn't really help my use case and on the understanding that it was probably going away anyway.

**Comment by jim-minter** [comment]
@stevekuznetsov `if/then` and `PATH` changes made and pushed FYI

File: lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb (hunk 7 above)
Comment (danmcp) [inline lib/vagrant-openshift/provisioner/configure_docker_server_linux.rb]: Then I don't any any issue with them existing in the workspace.

File: lib/vagrant-openshift/action/download_artifacts_sti.rb (hunk 2 above)
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/download_artifacts_sti.rb]: A function `is_windows?` or a comment, so someone reading thing knows why this statement is here

File: lib/vagrant-openshift/helper/command_helper.rb (hunk 4 above)
Comment (stevekuznetsov) [inline lib/vagrant-openshift/helper/command_helper.rb]: Sounds good to me

**Comment by jim-minter** [comment]
@stevekuznetsov ping, would it be possible to get feedback on next steps required for this soon?

**Comment by jim-minter** [comment]
@stevekuznetsov sorry, my ping crossed with you reviewing the PR :)

All changes pushed, including the comment you just requested.

File: lib/vagrant-openshift/hooks/configure_aws.rb [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    module Hooks
+      class ConfigureAWS
+        def initialize(app, env)
+          @app = app
+        end
+
+        def call(env)
+          if env[:machine].config.openshift.autoconfigure_aws == true
+            aws_creds = Vagrant::Openshift::Aws::aws_creds()
+            raise "AWS credentials not configured" if aws_creds.nil?
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: Are we ever going to be able to handle this outside of  `aws_creds()`? If not, makes sense to `raise` the error inside the function and save each caller the hassle.

File: lib/vagrant-openshift/hooks/configure_aws.rb [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+module Vagrant
+  module Openshift
+    module Hooks
+      class ConfigureAWS
+        def initialize(app, env)
+          @app = app
+        end
+
+        def call(env)
+          if env[:machine].config.openshift.autoconfigure_aws == true
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: Don't need the literal comparison to a `bool`.
```rb
if env[:machine].config.openshift.autoconfigure_aws
  ...
end
```

File: lib/vagrant-openshift/hooks/configure_aws.rb [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            aws_creds = Vagrant::Openshift::Aws::aws_creds()
+            raise "AWS credentials not configured" if aws_creds.nil?
+
+            aws = env[:machine].config.vm.get_provider_config(:aws)
+            if !aws.nil?
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: I'm not following this logic -- if we are asked to auto-configure AWS settings, why would this be `nil`, and, if it is, and we don't set up the secrets, won't everything fail when we start making AWS API calls in the future?

File: lib/vagrant-openshift/hooks/configure_aws.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+              aws.keypair_name = aws_creds["AWSKeyPairName"]
+
+              box_info = YAML.load_file(Pathname.new(File.expand_path("#{__FILE__}/../../templates/command/init-openshift/box_info.yaml")))
+              if !aws.ami.start_with?("ami-")
+                os, stage = aws.ami.split(":", 2)
+                env[:machine].config.ssh.username = box_info[os.to_sym][stage.to_sym][:aws][:ssh_user]
+              end
+            end
+            env[:machine].config.ssh.private_key_path = [aws_creds["AWSPrivateKeyPath"]]
+            env[:machine].config.vm.box = "https://github.com/mitchellh/vagrant-aws/raw/master/dummy.box"
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: Doesn't `vagrant up --provider=aws` set this, the SSH key, etc, for us?

File: lib/vagrant-openshift/hooks/windows_wait.rb [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+module Vagrant
+  module Openshift
+    module Hooks
+      class WindowsWait
+        def initialize(app, env)
+          @app = app
+        end
+
+        def call(env)
+          if env[:machine].config.openshift.autoconfigure_aws == true && env[:machine].config.ssh.username == "Administrator"
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/hooks/windows_wait.rb]: Since the `env[:machine].config.ssh.username == "Administrator"` block happens often, I think it would be nice to see it as a explanatory helper function:
```rb
def is_windows?(machine)
  return machine.config.ssh.username == "Administrator"
end
```

File: lib/vagrant-openshift/action/run_sti_tests.rb [hunk 14]
```diff
@@ -34,7 +34,7 @@ def call(env)
           cmds = ['OUTPUT_COVERAGE=/tmp/sti/artifacts/coverage hack/test-go.sh']
 
           if @options[:all]
-            cmds << 'STI_TIMEOUT="--timeout 240s" hack/test-integration.sh'
```
Comment (jim-minter) [inline lib/vagrant-openshift/action/run_sti_tests.rb]: STI_TIMEOUT was renamed in source-to-image some time ago and this was missed, so it was a no-op.  S2I_TIMEOUT upstream is set by default to a sufficiently similar value (and things have been working fine anyway) that I figured removal was the best option.

File: lib/vagrant-openshift/action/build_sti_base_windows.rb [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+# Create Administrator's home directory
+Out-Null | &bash --login
+&bash -c "mkdir /home/Administrator/.ssh; curl -o /home/Administrator/.ssh/authorized_keys http://169.254.169.254/latest/meta-data/public-keys/0/openssh-key"
+&bash -c "mkdir /data"
+$stream = New-Object System.IO.StreamWriter @("c:\cygwin\home\Administrator\.bash_profile", $true)
+$stream.Write("export GOPATH='C:\cygwin\data'`n")
+$stream.Close()
+
+# Download and install Docker client
+$client.DownloadFile("https://get.docker.com/builds/Windows/x86_64/docker-1.10.3.exe", "C:\cygwin\usr\local\bin\docker.exe")
```
Comment (jim-minter) [inline lib/vagrant-openshift/action/build_sti_base_windows.rb]: I don't regard it as a problem and I'm not aware of any obvious alternatives anyway.

File: lib/vagrant-openshift/action/build_sti_base_windows.rb [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+          images.sort_by! {|image| image.name[-10,10]}
+          @env[:ui].info("Using base AMI #{images.last.name}, #{images.last.id}")
+
+          @env[:ui].info("Creating instance...")
+          instance = compute.servers.create(
+            :tags => {"Name" => "#{@options[:instance_prefix]}-windows2012r2_#{Time.now.to_i}"},
+            :image_id => images.last.id,
+            :flavor_id => "m4.large",
+            :key_name => "libra",
+            :subnet_id => "subnet-cf57c596",
```
Comment (jim-minter) [inline lib/vagrant-openshift/action/build_sti_base_windows.rb]: Hmm, I don't think any of these values are derivable elsewhere - I think the alternative is to have an option for each (flavor_id, key_name, subnet_id). I can do this, it just means that whenever an AMI is built, that information has to be to hand somewhere else. If I add the options, would you accept defaults set as above, or not? (straight question).

File: lib/vagrant-openshift/action/generate_template.rb [hunk 17]
```diff
@@ -48,10 +49,13 @@ def call(env)
           @openstack_creds_file = Pathname.new(File.expand_path(@openstack_creds_file))
           box_info[:openstack_creds_file] = @openstack_creds_file
 
-          @aws_creds_file = ENV['AWS_CREDS'].nil? || ENV['AWS_CREDS'] == '' ? "~/.awscred" : ENV['AWS_CREDS']
-          @aws_creds_file = Pathname.new(File.expand_path(@aws_creds_file))
-          box_info[:aws_creds_file] = @aws_creds_file
-          find_ami_from_tag(box_info)
+          if !box_info[:aws].nil? && !box_info[:aws][:ami_tag_prefix].nil?
+            aws_creds = Vagrant::Openshift::Aws::aws_creds()
```
Comment (jim-minter) [inline lib/vagrant-openshift/action/generate_template.rb]: done

File: lib/vagrant-openshift/hooks/configure_aws.rb (hunk 10 above)
Comment (jim-minter) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: done

File: lib/vagrant-openshift/hooks/configure_aws.rb (hunk 9 above)
Comment (jim-minter) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: good point - done, will be in next push

File: lib/vagrant-openshift/hooks/configure_aws.rb (hunk 11 above)
Comment (jim-minter) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: ok, so will raise an exception if aws is nil

File: lib/vagrant-openshift/hooks/configure_aws.rb (hunk 12 above)
Comment (jim-minter) [inline lib/vagrant-openshift/hooks/configure_aws.rb]: No, it doesn't - origin's current Vagrantfile is a good example of that.

File: lib/vagrant-openshift/action/build_sti_base_windows.rb (hunk 16 above)
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/build_sti_base_windows.rb]: No, I don't think referring to a specific key name or subnet object in AWS in the code for this repo is appropriate. We've always encoded that information in external config for that reason before.

**Comment by jim-minter** [comment]
@stevekuznetsov all changes pushed except around the hard-coding of things like ':flavor_id => "m4.large"', on which I'm waiting for further comment from you.

It's my end of day now; I'd appreciate if you'd be able to provide any final actionable comments today your time so that we can get this closed off tomorrow?  Thanks.

**Comment by stevekuznetsov** [comment]
> It's my end of day now; I'd appreciate if you'd be able to provide any final actionable comments today your time so that we can get this closed off tomorrow? Thanks.

Can do, I'll do a final review now, nothing major was left on the last one. We should probably also [test] this

**Comment by stevekuznetsov** [comment]
@bparees did you want to weigh in on this?

**Comment by bparees** [comment]
> @bparees did you want to weigh in on this?

based on our conversation, no :)

**Comment by jim-minter** [comment]
@stevekuznetsov fixes to hard-coding of subnet_id etc. pushed.  NB FYI - the same subnet_id is also currently hard-coded in the origin Vagrantfile.

**Comment by stevekuznetsov** [comment]
Changes look good to me. Let's let the tests run their course overnight and we can merge this in tomorrow.

File: lib/vagrant-openshift/action/generate_template.rb [hunk 18]
```diff
@@ -103,33 +108,6 @@ def call(env)
 
           @app.call(env)
         end
-
-        private
-
-        def find_ami_from_tag(box_info)
-          return if box_info[:aws].nil? || box_info[:aws][:ami_tag_prefix].nil?
-          @env[:ui].info("Reading AWS credentials from #{@aws_creds_file.to_s}")
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/generate_template.rb]: New functions missing this line.

File: lib/vagrant-openshift/action/generate_template.rb [hunk 19]
```diff
@@ -103,33 +108,6 @@ def call(env)
...
-
-            fog_config = {
-                :provider              => :aws,
-                :region                => box_info[:aws][:ami_region],
-                :aws_access_key_id     => aws_creds['AWSAccessKeyId'],
-                :aws_secret_access_key => aws_creds['AWSSecretKey'],
-            }
-
-            aws_compute = Fog::Compute.new(fog_config)
-            @env[:ui].info("Searching for latest base AMI")
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/generate_template.rb]: New functions missing this line.

File: lib/vagrant-openshift/action/generate_template.rb [hunk 20]
```diff
@@ -103,33 +108,6 @@ def call(env)
...
-            image_filter = {'Owner' => 'self', 'name' => "#{box_info[:aws][:ami_tag_prefix]}*", 'state' => 'available' }
-            image_filter['tag:Name'] = @options[:required_name_tag] unless @options[:required_name_tag].nil?
-            images = aws_compute.images.all(image_filter)
-            latest_image = images.sort_by{ |i| i.name.split("_")[-1].to_i }.last
-            box_info[:aws][:ami] = latest_image.id
-            @env[:ui].info("Found: #{latest_image.id} (#{latest_image.name})")
```
Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/generate_template.rb]: New functions missing this line.

Comment (stevekuznetsov) [inline lib/vagrant-openshift/action/generate_template.rb]: Would suggest changes like https://github.com/stevekuznetsov/vagrant-openshift/commit/4b47e84c6ff12b78216d14cb59da0e298c465553 here
